mypy/nodes.py

```diff
@@ -1971,7 +1975,23 @@ class is generic then it will be a type constructor of higher kind.
     subtypes = None  # type: Set[TypeInfo] # Direct subclasses encountered so far
     names = None  # type: SymbolTable      # Names defined directly in this type
     is_abstract = False                    # Does the class have any abstract attributes?
+    is_protocol = False                    # Is this a protocol class?
+    runtime_protocol = False               # Does this protocol support isinstance checks?
     abstract_attributes = None  # type: List[str]
+    protocol_members = None  # type: List[str]
+
+    # These represent structural subtype matrices. Note that these are shared
+    # by all 'Instance's with given TypeInfo.
+    # If concurrent/parallel type checking will be added in future,
+    # then there should be one matrix per thread/process to avoid false negatives
+    # during the type checking phase.
```

Comment:
Extend this comment explaining _why_ we use ``assuming`` and ``inferring`` attributes on ``TypeInfo`` instead of passing them as kwargs to ``is_subtype``, ``infer_constraint``, etc.
Also explain why we can't use sets here (we need to use ``is_same_type`` instead of relying on hash to correctly treat unions).